cpp/src/arrow/testing/gtest_util.h

```diff
@@ -121,6 +121,17 @@ class Result;
   ASSERT_OK_AND_ASSIGN_IMPL(ARROW_ASSIGN_OR_RAISE_NAME(_error_or_value, __COUNTER__), \
                             lhs, rexpr);
 
+#define EXPECT_OK_AND_ASSIGN_IMPL(status_name, lhs, rexpr) \
+  auto status_name = (rexpr);                              \
+  ARROW_EXPECT_OK(status_name.status());                   \
+  if (status_name.ok()) {                                  \
+    lhs = std::move(status_name).ValueOrDie();             \
+  }
+
+#define EXPECT_OK_AND_ASSIGN(lhs, rexpr)                                              \
```

Comment:
When you rebase, maybe implement this with ASSIGN_OR_HANDLE_ERROR_IMPL?

Reply:
I'll take a look.